### PR TITLE
Adding Page Shield support

### PR DESCRIPTION
**Pull request by wiremanb**
## Description
Added support for Page Shield based on developer docs.

## Has your change been tested?
All changes have been tested with test suites. Also, have tested using a local driver application 

## Screenshots (if appropriate):
Screenshot 2023-12-05 at 7 05 17 PM

## Types of changes

What sort of change does your code introduce/modify?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] This change is using publicly documented in [cloudflare/api-schemas](https://github.com/cloudflare/api-schemas) 
      and relies on stable APIs.

[1]: https://help.github.com/articles/closing-issues-using-keywords/

### CODE REVIEW

**Comment by wiremanb** [comment]
@jacobbednarz do you think it is necessary to use ptr's for boolean values in this? How stringent do you want me to be with the linter issues?

**Comment by jacobbednarz** [comment]
the semgrep rules are in place from past issues we've seen with usage (especially the terraform provider). i'm not sure why but semgrep should be pointing you at https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#booleans which explains it a little more too.

**Comment by wiremanb** [comment]
I noticed some other files that don't use it, that's why I ask. I will update and push another. Apologies for the repetitive nature.

**Comment by jacobbednarz** [comment]
i've gone through and applied a few updates from https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods to get this closer to the library conventions and make our future maintenance a bit easier :)

**Comment by wiremanb** [comment]
> i've gone through and applied a few updates from https://github.com/cloudflare/cloudflare-go/blob/master/docs/conventions.md#methods to get this closer to the library conventions and make our future maintenance a bit easier :)

Thank you! I just noticed. I don't do a lot of public contribution.. thanks for helping me out :-)

File: page_shield_connections.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/goccy/go-json"
+)
+
+// ListPageShieldConnectionsParams represents parameters for a page shield connection request.
+type ListPageShieldConnectionsParams struct {
```
Comment (jacobbednarz) [inline page_shield_connections.go:12]: for these params, are they expected to be URL query parameters if passed in? if so, we will need to use the `url` struct tags to handle converting them for us. examples: https://github.com/cloudflare/cloudflare-go/blob/0904ebc0820ca6aa225455154237328e34c67754/dns.go#L56-L66

File: page_shield_connections_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	"context"
+	"fmt"
+	"net/http"
+	"testing"
+
+	"github.com/goccy/go-json"
+	"github.com/stretchr/testify/assert"
+)
+
+// Mock data for PageShieldConnections
```
Comment (jacobbednarz) [inline page_shield_connections_test.go]:
```suggestion
// Mock data for PageShieldConnections.
```

File: page_shield_scripts.go [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+	PageURLs                []string `json:"page_urls"`
+	URL                     string   `json:"url"`
+	URLContainsCdnCgiPath   *bool    `json:"url_contains_cdn_cgi_path,omitempty"`
+}
+
+// ListPageShieldScriptsParams represents a PageShield Script request parameters.
+//
+// API reference: https://developers.cloudflare.com/api/operations/page-shield-list-page-shield-scripts#Query-Parameters
+type ListPageShieldScriptsParams struct {
+	Direction           string `json:"direction"`
```
Comment (jacobbednarz) [inline page_shield_scripts.go]: similar to the above on query parameters.

File: page_shield_test.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+import (
+	"context"
+	"net/http"
+	"testing"
+
+	"github.com/goccy/go-json"
+	"github.com/stretchr/testify/assert"
+)
+
+// Mock PageShieldSettings data
```
Comment (jacobbednarz) [inline page_shield_test.go]:
```suggestion
// Mock PageShieldSettings data.
```

File: page_shield_connections.go (hunk 1 above)
Comment (wiremanb) [reply page_shield_connections.go:12]: url params

File: page_shield_scripts.go (hunk 3 above)
Comment (wiremanb) [reply page_shield_scripts.go]: same here: url.

**Comment by jacobbednarz** [comment]
it looks like the force push has lost my intermediary commits to get the conventions and linter happy.

**Comment by wiremanb** [comment]
> it looks like the force push has lost my intermediary commits to get the conventions and linter happy.

aaanndd that's why they say don't do a force push... I have your changes in the PR checked out on my machine. Do you want me to close this PR and open a new one?

**Comment by jacobbednarz** [comment]
> aaanndd that's why they say don't do a force push...

😂 check out https://blog.developer.atlassian.com/force-with-lease/

> I have your changes in the PR checked out on my machine. Do you want me to close this PR and open a new one?

do you just want to cherry-pick them onto this one? whatever is easiest for you to do.

**Comment by wiremanb** [comment]
To be honest, I have not cherry-picked before. That might cause more issues. I suppose I could try it and then if I screw it up, I have a local tarball of the changes.. I could just try again :-D

File: page_shield_connections_test.go (hunk 2 above)
Comment (jacobbednarz) [inline page_shield_connections_test.go]:
```suggestion
// Mock data for PageShieldConnections.
```

File: page_shield_test.go (hunk 4 above)
Comment (jacobbednarz) [inline page_shield_test.go]:
```suggestion
// Mock PageShieldSettings data.
```

**Comment by wiremanb** [comment]
@jacobbednarz thanks for the help with this. Seriously, appreciate your patience. I royally screwed this up on the send.

**Comment by jacobbednarz** [comment]
no worries! thanks for the contribution 🏅

**Pull request merged**